### PR TITLE
GH-15809: fixes loglikelihood and aic for glm generic model

### PR DESCRIPTION
**Pull request by syzonyuliia**
#15809

### CODE REVIEW

**Review by wendycwong — CHANGES_REQUESTED** [review]

File: h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py [hunk 1]
```diff
@@ -30,18 +34,40 @@ def test(x, y, output_test, strip_part, algo_name, generic_algo_name, family):
...
+        print(glm._model_json["output"]["training_metrics"])
+        print("glm calc like training metrics:")
+        print(glm_calc_like._model_json["output"]["training_metrics"])
+        print("metrics:")
+        print(metrics)
+        
+        assert math.isclose(glm_calc_like._model_json["output"]["training_metrics"]._metric_json["AIC"], 
+                            metrics._metric_json["AIC"], rel_tol=1e-3), "The numbers are not close enough."
+        assert math.isclose(-glm_calc_like._model_json["output"]["training_metrics"]._metric_json["loglikelihood"], 
+                            metrics._metric_json["loglikelihood"], rel_tol=1e-3), "The numbers are not close enough."
```
Comment (wendycwong) [inline h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py:69]: Julia:  I was surpised that the results are a little different.  I was hoping they should match exactly....  Do you know the reason for the discrepancies?

File: h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py [hunk 2]
```diff
@@ -62,7 +88,7 @@ def mojo_model_test_ordinal():
     test(["Origin", "Distance", "IsDepDelayed"], "fDayOfWeek", compare_output, 'GLM Model: summary',
          'ModelMetricsOrdinalGLM: glm',
          'ModelMetricsOrdinalGLMGeneric: generic', 'ordinal')
-    
+
 
 pyunit_utils.run_tests([
```
Comment (wendycwong) [inline h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py:93]: if the ordinal is giving you problem, just comment it out for now.

File: h2o-algos/src/main/java/hex/generic/GenericModel.java [hunk 3]
```diff
@@ -131,6 +154,34 @@ protected PredictScoreResult predictScoreImpl(Frame fr, Frame adaptFrm, String d
             return predictScoreMojoImpl(fr, destination_key, j, computeMetrics);
         } else
             return super.predictScoreImpl(fr, adaptFrm, destination_key, j, computeMetrics, customMetricFunc);
+//            return super.predictScoreImpl(fr, adaptFrm, destination_key, j, true, customMetricFunc);
```
Comment (tomasfryda) [inline h2o-algos/src/main/java/hex/generic/GenericModel.java:157]:
```suggestion
```

File: h2o-algos/src/main/java/hex/generic/GenericModel.java [hunk 4]
```diff
@@ -131,6 +154,34 @@ protected PredictScoreResult predictScoreImpl(Frame fr, Frame adaptFrm, String d
...
+        if (!_algoName.equals("glm")) {
+            return 0;
+        } else {
+            long betasCount = Arrays.stream(((GlmMojoModelBase) this.genModel()).getBeta()).filter(b -> b != 0).count();
+            return -2 * likelihood + 2 * betasCount;
+        }
+    }
+
+    @Override
+    public double likelihood(double w, double y, double[] f) {
```
Comment (tomasfryda) [inline h2o-algos/src/main/java/hex/generic/GenericModel.java:177]: If it's negative log likelihood, I'd prefer to name it `negative_log_likelihood` or something like  that.

File: h2o-bindings/bin/gen_R.py [hunk 5]
```diff
@@ -284,6 +284,8 @@ def get_help(param, indent=0):
     phelp = param.get('help')
     if not phelp:
         return
+    # fix braces so that CRAN doesn't complain
+    phelp = phelp.replace("{", "\{").replace("}", "\}")
     if ptype == 'boolean':
```
Comment (tomasfryda) [inline h2o-bindings/bin/gen_R.py:289]: This should already be in both master and rel-3.44.0. https://github.com/h2oai/h2o-3/pull/16010/files#diff-231f2d963f774362ad0fb01f7825ddc8cf1f32d59ce0c5dd053abea9402b055fR287-R288

I'm confused why I see the change here. Current master already has that (https://github.com/h2oai/h2o-3/blob/master/h2o-bindings/bin/gen_R.py#L287-L288),

File: h2o-core/src/main/java/water/api/schemas3/ModelMetricsBinomialV3.java [hunk 6]
```diff
@@ -18,6 +18,12 @@ public class ModelMetricsBinomialV3<I extends ModelMetricsBinomial, S extends Mo
   @API(help="The logarithmic loss for this scoring run.", direction=API.Direction.OUTPUT)
   public double logloss;
 
+  @API(help="The negative logarithmic likelihood for this scoring run.", direction=API.Direction.OUTPUT)
+  public double loglikelihood;
```
[the remark this answers is not in the document]
Comment (syzonyuliia) [reply h2o-core/src/main/java/water/api/schemas3/ModelMetricsBinomialV3.java:22]: Right, thank you, that's correct, I'll remove "negative"

File: h2o-algos/src/main/java/hex/generic/GenericModel.java (hunk 4 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/generic/GenericModel.java:176]: This is a confusing issue.  We have been using likelihood to mean negative_log_likelihood.  There will be so many changes.  For now I suggest just add a comment indicating that this is actually negative log likelihood function.  Customer is waiting for this one.  Moving forward, I think I want to change it to negative_log_likelihood function.  Here is the issue: https://github.com/h2oai/h2o-3/issues/16059

File: h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py [hunk 7]
```diff
@@ -14,8 +16,10 @@ def test(x, y, output_test, strip_part, algo_name, generic_algo_name, family):
 
     # GLM
     airlines = h2o.import_file(path=pyunit_utils.locate("smalldata/testng/airlines_train.csv"))
-    glm = H2OGeneralizedLinearEstimator(nfolds = 2, family = family, max_iterations=2) # alpha = 1, lambda_ = 1, bad values, use default
-    glm.train(x = x, y = y, training_frame=airlines, validation_frame=airlines, )
+    glm = H2OGeneralizedLinearEstimator(nfolds=2, family=family, max_iterations=2, 
+                                        compute_p_values=(family == "gaussian"), 
+                                        remove_collinear_columns=(family == "gaussian")) # alpha = 1, lambda_ = 1, bad values, use default
```
Comment (wendycwong) [inline h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py:21]: add

seed=12345

to the model building

File: h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py [hunk 8]
```diff
@@ -30,18 +34,40 @@ def test(x, y, output_test, strip_part, algo_name, generic_algo_name, family):
...
 
+    print(generic_mojo_model_from_file._model_json["output"]["training_metrics"])
     generic_mojo_filename = tempfile.mkdtemp("zip", "genericMojo");
     generic_mojo_filename = generic_mojo_model_from_file.download_mojo(path=generic_mojo_filename)
     assert os.path.getsize(generic_mojo_filename) == os.path.getsize(original_model_filename)
 
+    if family != 'ordinal':  # loglikelihood calculation not available for ordinal family yet
+        glm_calc_like = H2OGeneralizedLinearEstimator(nfolds=2, family=family, max_iterations=2, calc_like=True,
+                                                      compute_p_values=(family == "gaussian"),
+                                                      remove_collinear_columns=(family == "gaussian"))
```
Comment (wendycwong) [inline h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py:56]: add 

seed=12345

to model building

File: h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py [hunk 9]
```diff
@@ -30,18 +34,40 @@ def test(x, y, output_test, strip_part, algo_name, generic_algo_name, family):
...
+        print(glm._model_json["output"]["training_metrics"])
+        print("glm calc like training metrics:")
+        print(glm_calc_like._model_json["output"]["training_metrics"])
+        print("metrics:")
+        print(metrics)
+        
+        assert math.isclose(glm_calc_like._model_json["output"]["training_metrics"]._metric_json["AIC"], 
+                            metrics._metric_json["AIC"], rel_tol=1e-3), "The numbers are not close enough."
+        assert math.isclose(-glm_calc_like._model_json["output"]["training_metrics"]._metric_json["loglikelihood"], 
+                            metrics._metric_json["loglikelihood"], rel_tol=1e-3), "The numbers are not close enough."
```
Comment (wendycwong) [inline h2o-py/tests/testdir_generic_model/pyunit_generic_model_mojo_glm.py:69]: change 

1e-3

to

1e-6

**Review by wendycwong — CHANGES_REQUESTED** [review]

File: h2o-py/h2o/model/metrics_base.py [hunk 10]
```diff
@@ -342,7 +349,8 @@ def loglikelihood(self):
         ...                validation_frame = valid)
         >>> pros_glm.loglikelihood()
         """
-        return self._metric_json['loglikelihood']
+        if self._algo == 'glm':
+            return self._metric_json['loglikelihood']
```
Comment (tomasfryda) [inline h2o-py/h2o/model/metrics_base.py:353]: I'd prefer if you'd not care about the algo. So I'd suggest to use `.get("loglikelihood")` instead of checking the `algo`.

```suggestion
        return self._metric_json.get('loglikelihood')
```

File: h2o-py/h2o/model/metrics_base.py [hunk 11]
```diff
@@ -319,7 +325,8 @@ def aic(self):
         ...                validation_frame = valid)
         >>> pros_glm.aic()
         """
-        return self._metric_json['AIC']
+        if self._algo == 'glm':
+            return self._metric_json['AIC']
```
Comment (tomasfryda) [inline h2o-py/h2o/model/metrics_base.py:329]: I'd prefer if you'd not care about the algo. So I'd suggest to use `.get("AIC")` instead of checking the `algo`. This way if we expose these metrics e.g. from the metalearner to stacked ensemble or if Wendy adds it to GAM we will not need to modify this part.

```suggestion
        return self._metric_json.get('AIC')
```

File: h2o-core/src/main/java/hex/ModelMetricsBinomial.java [hunk 12]
```diff
@@ -19,15 +19,19 @@
 public class ModelMetricsBinomial extends ModelMetricsSupervised {
   public final AUC2 _auc;
   public final double _logloss;
+  public final double _loglikelihood;
+  public final double _aic;
   public double _mean_per_class_error;
   public final GainsLift _gainsLift;
 
   public ModelMetricsBinomial(Model model, Frame frame, long nobs, double mse, String[] domain,
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsBinomial.java:27]: I would probably prefer adding one more constructor here and revert all the changes for non-GLM algos that don't support likelihood.

```java
public ModelMetricsBinomial(Model model, Frame frame, long nobs, double mse, String[] domain,
                              double sigma, AUC2 auc, double logloss, double loglikelihood, double aic, GainsLift gainsLift,
                              CustomMetric customMetric) {
....
}

public ModelMetricsBinomial(Model model, Frame frame, long nobs, double mse, String[] domain,
                              double sigma, AUC2 auc, double logloss, GainsLift gainsLift, CustomMetric customMetric) {
    this(model, frame, nobs, mse, domain, sigma, auc, logloss, Double.NaN, Double.NaN, gainsLift, customMetric);
}
```

This way you wouldn't need to modify all the constructors in all the models that use `ModelMetricsBinomial` and you could provide some reasonable values as defaults for loglikelihood and AIC. For example if I add a new algo, I want it to be easy so I don't have to think with which values I should fill fields that the algo doesn't support. Having the defaults in a new constructor makes it easier and more consistent.

I would recommend using some value like `NaN` for the default - if the loglikelihood is really log likelihood (and not negative log likelihood) then `0` is the maximum and that doesn't seem like a good default (e.g. think about sorting by that in the leaderboard etc). (And similarly for AIC.)

File: h2o-core/src/main/java/hex/ModelMetricsMultinomial.java [hunk 13]
```diff
@@ -18,14 +18,20 @@ public class ModelMetricsMultinomial extends ModelMetricsSupervised {
   public final float[] _hit_ratios;         // Hit ratios
   public final ConfusionMatrix _cm;
   public final double _logloss;
+  public final double _loglikelihood;
+  public final double _aic;
   public double _mean_per_class_error;
   public MultinomialAUC _auc;
   
-  public ModelMetricsMultinomial(Model model, Frame frame, long nobs, double mse, String[] domain, double sigma, ConfusionMatrix cm, float[] hr, double logloss, MultinomialAUC auc, CustomMetric customMetric) {
+  public ModelMetricsMultinomial(Model model, Frame frame, long nobs, double mse, String[] domain, double sigma,
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsMultinomial.java:26]: Please see suggestion: https://github.com/h2oai/h2o-3/pull/16025/files#r1481117346

File: h2o-core/src/main/java/hex/ModelMetricsRegression.java [hunk 14]
```diff
@@ -13,21 +13,28 @@
...
+  public double loglikelihood() { return _loglikelihood; }
+  public double aic() { return _AIC; }
   @SuppressWarnings("unused")
   public double mean_residual_deviance() { return _mean_residual_deviance; }
   public final double _mean_absolute_error;
   public double mae() { return _mean_absolute_error; }
   public final double _root_mean_squared_log_error;
   public double rmsle() { return _root_mean_squared_log_error; }
-  public ModelMetricsRegression(Model model, Frame frame, long nobs, double mse, double sigma, double mae,double rmsle, double meanResidualDeviance, CustomMetric customMetric) {
+  public ModelMetricsRegression(Model model, Frame frame, long nobs, double mse, double sigma, double mae,double rmsle,
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsRegression.java:30]: Please see suggestion: https://github.com/h2oai/h2o-3/pull/16025/files#r1481117346

File: h2o-core/src/main/java/hex/ModelMetricsBinomial.java [hunk 15]
```diff
@@ -197,6 +206,11 @@ public static class MetricBuilderBinomial<T extends MetricBuilderBinomial<T>> ex
         // Compute log loss
         _logloss += w * MathUtils.logloss(err);
       }
+
+      if(m != null && m.getClass().toString().contains("Generic")) {
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsBinomial.java:210]: This looks weird to me. I would prefer not to hardwire some behavior based on the class name.

Would it make sense to add a new predicate e.g. `supportsLikelihood()` so you communicate the intent of the condition. I am unsure why Generic models should be different than others. Or I guess you could use something like `m instanceof GenericModel` but I'd prefer if you consider some alternative approach that would be more in line with OOP.

File: h2o-core/src/main/java/hex/ModelMetricsBinomial.java [hunk 16]
```diff
@@ -256,18 +271,24 @@ private ModelMetrics makeModelMetrics(final Model m, final Frame f, final Frame
...
+      double loglikelihood = Double.NaN;
+      double aic = Double.NaN;
       double logloss = Double.NaN;
       double sigma = Double.NaN;
       final AUC2 auc;
       if (_wcount > 0) {
         sigma = weightedSigma();
         mse = _sumsqe / _wcount;
         logloss = _logloss / _wcount;
+        if(m != null && m.getClass().toString().contains("Generic")) {
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsBinomial.java:283]: https://github.com/h2oai/h2o-3/pull/16025/files#r1481273648

File: h2o-core/src/main/java/hex/ModelMetricsMultinomial.java [hunk 17]
```diff
@@ -302,6 +313,11 @@ public MetricBuilderMultinomial( int nclasses, String[] domain, MultinomialAucTy
       if(_calculateAuc) {
         calculateAucsPerRow(ds, iact, w);
       }
+
+
+      if(m != null && m.getClass().toString().contains("Generic")) {
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsMultinomial.java:318]: https://github.com/h2oai/h2o-3/pull/16025/files#r1481273648

File: h2o-core/src/main/java/hex/ModelMetricsMultinomial.java [hunk 18]
```diff
@@ -360,10 +379,14 @@ private void calculateAucsPerRow(double ds[], int iact, double w){
         }
         mse = _sumsqe / _wcount;
         logloss = _logloss / _wcount;
+        if(m != null && m.getClass().toString().contains("Generic")) {
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsMultinomial.java:382]: https://github.com/h2oai/h2o-3/pull/16025/files#r1481273648

File: h2o-core/src/main/java/hex/ModelMetricsRegression.java [hunk 19]
```diff
@@ -147,6 +157,10 @@ public MetricBuilderRegression(Distribution dist) {
           _sumdeviance += _dist.deviance(w, yact[0], ds[0]);
         }
       }
+
+      if(m != null && m.getClass().toString().contains("Generic")) {
```
Comment (tomasfryda) [inline h2o-core/src/main/java/hex/ModelMetricsRegression.java:161]: https://github.com/h2oai/h2o-3/pull/16025/files#r1481273648

File: h2o-py/h2o/model/metrics_base.py (hunk 10 above)
Comment (wendycwong) [reply h2o-py/h2o/model/metrics_base.py]: I removed the algorithm check per your suggestion.

File: h2o-py/h2o/model/metrics_base.py (hunk 11 above)
Comment (wendycwong) [reply h2o-py/h2o/model/metrics_base.py]: Removed the algo check per your suggestions.

**Review by wendycwong — APPROVED** [review]
